File: flutter_app/test/unit/application/use_cases/create_todo_use_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:habit_todo_app/core/application/use_cases/create_todo_use_case.dart';
import 'package:habit_todo_app/core/domain/entities/todo_entity.dart';
import 'package:habit_todo_app/core/domain/entities/todo_icon_type.dart';
import 'package:habit_todo_app/core/domain/repositories/todo_repository.dart';
import 'package:mocktail/mocktail.dart';

class MockTodoRepository extends Mock implements TodoRepository {}

void main() {
  setUpAll(() {
    registerFallbackValue(TodoEntity(
      id: 'fallback',
      title: 'Fallback',
      createdAt: DateTime(2024, 1, 1),
    ));
  });

  group('CreateTodoUseCase', () {
    late CreateTodoUseCase useCase;
    late MockTodoRepository mockRepository;

    setUp(() {
      mockRepository = MockTodoRepository();
      useCase = CreateTodoUseCase(mockRepository);
    });

    test('should create todo with valid input', () async {
      // Arrange
      final expectedTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      final result = await useCase.execute(title: 'Test Todo');

      // Assert
      expect(result, equals(expectedTodo));
      verify(() => mockRepository.create(any())).called(1);
    });

    test('should create todo with all fields', () async {
      // Arrange
      final deadline = DateTime(2024, 1, 15, 20, 30);
      final expectedTodo = TodoEntity(
        id: 'todo_1',
        title: 'Complete Task',
        description: 'Description',
        deadline: deadline,
        iconType: TodoIconType.writing,
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      final result = await useCase.execute(
        title: 'Complete Task',
        description: 'Description',
        deadline: deadline,
        iconType: TodoIconType.writing,
      );

      // Assert
      expect(result, equals(expectedTodo));
      verify(() => mockRepository.create(any())).called(1);
    });

    test('should throw ArgumentError if title is empty', () async {
      // Act & Assert
      expect(
        () => useCase.execute(title: ''),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('should throw ArgumentError if title is only whitespace', () async {
      // Act & Assert
      expect(
        () => useCase.execute(title: '   '),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('should trim title and description', () async {
      // Arrange
      final expectedTodo = TodoEntity(
        id: 'todo_1',
        title: 'Trimmed Title',
        description: 'Trimmed Description',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      await useCase.execute(
        title: '  Trimmed Title  ',
        description: '  Trimmed Description  ',
      );

      // Assert
      verify(() => mockRepository.create(any(
              that: predicate<TodoEntity>(
            (todo) =>
                todo.title == 'Trimmed Title' &&
                todo.description == 'Trimmed Description',
          )))).called(1);
    });

    test('should generate ID if not provided', () async {
      // Arrange
      final expectedTodo = TodoEntity(
        id: 'generated_id',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      await useCase.execute(title: 'Test Todo');

      // Assert
      verify(() => mockRepository.create(any(
              that: predicate<TodoEntity>(
            (todo) => todo.id.isNotEmpty,
          )))).called(1);
    });

    test('should use provided ID if given', () async {
      // Arrange
      final expectedTodo = TodoEntity(
        id: 'custom_id',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      await useCase.execute(title: 'Test Todo', id: 'custom_id');

      // Assert
      verify(() => mockRepository.create(any(
              that: predicate<TodoEntity>(
            (todo) => todo.id == 'custom_id',
          )))).called(1);
    });

    test('should use default icon type if not provided', () async {
      // Arrange
      final expectedTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        iconType: TodoIconType.defaultIcon,
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.create(any()))
          .thenAnswer((_) async => expectedTodo);

      // Act
      await useCase.execute(title: 'Test Todo');

      // Assert
      verify(() => mockRepository.create(any(
              that: predicate<TodoEntity>(
            (todo) => todo.iconType == TodoIconType.defaultIcon,
          )))).called(1);
    });
  });
}

File: apps/web/test/unit/application/use_cases/create_todo_use_case.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { CreateTodoUseCase } from "../../../../src/core/application/use_cases/create_todo_use_case";
import type { TodoEntity } from "../../../../src/core/domain/entities/todo_entity";
import type { TodoRepository } from "../../../../src/core/domain/repositories/todo_repository";
import { Priority } from "../../../../src/core/domain/value_objects/priority";

describe("CreateTodoUseCase", () => {
  let mockRepository: TodoRepository;
  let useCase: CreateTodoUseCase;

  beforeEach(() => {
    mockRepository = {
      findAll: vi.fn(),
      findById: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
    };
    useCase = new CreateTodoUseCase(mockRepository);
  });

  it("should create todo with valid input", async () => {
    const mockTodo: TodoEntity = {
      id: "test-id",
      title: "Test Todo",
      description: "Test Description",
      priority: Priority.MEDIUM,
      completed: false,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    vi.mocked(mockRepository.create).mockResolvedValue(mockTodo);

    const result = await useCase.execute({
      title: "Test Todo",
      description: "Test Description",
      priority: Priority.MEDIUM,
    });

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.title).toBe("Test Todo");
      expect(mockRepository.create).toHaveBeenCalled();
    }
  });

  it("should fail when title is empty", async () => {
    const result = await useCase.execute({ title: "" });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Title is required");
    }
    expect(mockRepository.create).not.toHaveBeenCalled();
  });

  it("should fail when title is too long", async () => {
    const longTitle = "a".repeat(201);
    const result = await useCase.execute({ title: longTitle });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Title must not exceed 200 characters");
    }
  });

  it("should fail when description is too long", async () => {
    const longDescription = "a".repeat(1001);
    const result = await useCase.execute({
      title: "Test",
      description: longDescription,
    });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Description must not exceed 1000 characters");
    }
  });

  it("should fail when due date is in the past", async () => {
    const pastDate = new Date("2020-01-01").toISOString();
    const result = await useCase.execute({
      title: "Test",
      dueDate: pastDate,
    });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Due date must be in the future");
    }
  });

  it("should use default priority when not provided", async () => {
    const mockTodo: TodoEntity = {
      id: "test-id",
      title: "Test Todo",
      priority: Priority.MEDIUM,
      completed: false,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    vi.mocked(mockRepository.create).mockResolvedValue(mockTodo);

    const result = await useCase.execute({ title: "Test Todo" });

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.priority).toBe(Priority.MEDIUM);
    }
  });
});
